### DISCUSSION TITLE
process.env.VERCEL_URL is undefined

### DISCUSSION THREAD

**Discussion by fabien-h** [vercel/next.js#15902] [Help] [answered]
I'm using `process.env.VERCEL_URL` in an email that I send when the user creates an account.

In dev mode on my local machine, it's transformed into `http://localhost:3000/`. But once deployed, it is `undefined`. That's a bit weird.

**Comment by paulogdm** [comment, marked as answer]
You need to add "VERCEL_URL" to your environment variables UI: https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables

**Reply by fabien-h** [reply]
Ohhh. Ok. Thank you for your answer :)

Since I saw "populated by the system" and it's already there in dev mode, I guessed that it was initialized without human intervention.

Why not having all those protected variables defined for all projects?

PS : really, I'm the first one who does not understand this functionnality! SHAME SHAME SHAME!

**Reply by paulogdm** [reply]
Because we have a limit in the number of secrets. Also, we don't want to optimize for cases where our users don't need the variables at all.

**Reply by dwwoelfel** [reply]
@paulogdm Is there any way to get the VERCEL_URL to populate without using the UI? 

I'm trying to create a template repo that people can clone and deploy to vercel and I'd like to find a way to populate the VERCEL_URL through a configuration file.

**Reply by paulogdm** [reply]
You can use: https://vercel.com/docs/more/deploy-button.

**Reply by BrunoQuaresma** [reply]
🤔 I have the same question... how VERCEL_URL can be manually set if this is dynamic for each preview deployment?

**Reply by paulogdm** [reply]
As a system variable, it will not be manually set by the user.
This is also possible:
![Screenshot_2021-02-07_09h00m](https://user-images.githubusercontent.com/7690649/107211603-8df47400-69e4-11eb-87f9-64e949ce4390.png)

**Reply by ulysse-qd** [reply]
This is what I have but it still doesn't work... It worked in the beginning but not anymore I don't know what happened

**Reply by RemyMachado** [reply]
> This is what I have but it still doesn't work... It worked in the beginning but not anymore I don't know what happened

EXACTLY!

It suddenly stops working for no reason.

@paulogdm do you have any info on this? I found another thread online with no response with this exact problem.

It used to work with `NEXT_PUBLIC_URL=https://$NEXT_PUBLIC_VERCEL_URL`, but now I'm just getting this lowercase output:
`https://$next_public_vercel_url`. This isn't getting interpolated anymore. 

I'm confused, it's like `dotenv-expand` was removed from vercel's internal packages haha.

**Comment by garbit** [comment]
Can I assign ```VERCEL_URL``` to another environment variable i.e. `AUTH_ORIGIN` = `$VERCEL_URL` in the UI?
